`option` and `optgroup` shouldn't be focusable in a menulist

Ahmad-S792 opened the pull request:
#### b9e0d14b62e69a753f91941a5beeb540c56c1858
<pre>
`option` and `optgroup` shouldn&apos;t be focusable in a menulist

<a href="https://bugs.webkit.org/show_bug.cgi?id=249627">https://bugs.webkit.org/show_bug.cgi?id=249627</a>
rdar://problem/103717395

Reviewed by Tim Nguyen.

This patch aligns WebKit with Blink / Chromium and Gecko / Firefox.

Merge: <a href="https://chromium.googlesource.com/chromium/blink/+/db4eb7807f1591c3beebe2e40a3c57d9e57a3c85">https://chromium.googlesource.com/chromium/blink/+/db4eb7807f1591c3beebe2e40a3c57d9e57a3c85</a>

This patch modifies WebKit behavior to match other browsers,
where `option` and `optgroup` elements are not focusable at all
when in menulist within `select` element irrespective of &apos;tabindex&apos;
attribute and simplify code as well.

* Source/WebCore/html/HTMLOptionElement.cpp:
(HTMLOptionElement::isFocusable): return &apos;false&apos; when in menulist
* Source/WebCore/html/HTMLOptGroupElement.cpp:
(HTMLOptGroupElement::isFocusable): Ditto
* LayoutTests/fast/forms/select/menulist-focusable-items.html: Add Test Case
* LayoutTests/fast/forms/select/menulist-focusable-items-expected.txt: Add Test Case Expectation
* LayoutTests/fast/forms/select/listbox-focusable-items.html: Add Test Case
* LayoutTests/fast/forms/select/listbox-focusable-items-expected.txt: Add Test Case Expectation
* LayoutTests/fast/events/resources/tabindex-focus-blur-all-iframe1.html: Rebaselined
* LayoutTests/fast/events/resources/tabindex-focus-blur-all-iframe2.html: Rebaselined
* LayoutTests/fast/events/resources/tabindex-focus-blur-all-frame1.html: Rebaselined
* LayoutTests/fast/events/resources/tabindex-focus-blur-all-frame2.html: Rebaselined
* LayoutTests/platform/ios/TestExpectations: Add &apos;ios&apos; Expectation for &apos;fast/forms/select/listbox-focusable-items.html&apos;

Canonical link: <a href="https://commits.webkit.org/264755@main">https://commits.webkit.org/264755@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69af73e54c8c3e62ae3d3ad63c3e5511f81a7cbe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8544 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8834 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9048 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10199 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/8571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8553 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10823 "Built successfully") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8798 "Failed to checkout and rebase branch from PR 14507") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/11438 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8690 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/9718 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7683 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/10355 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/7018 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/7813 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/10355 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/8137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/7962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/10355 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/8439 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/16/builds/8798 "Failed to checkout and rebase branch from PR 14507") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/7716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/7724 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/11925 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1008 "Built successfully and passed tests") | [❌ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8181 "Failed to checkout and rebase branch from PR 14507") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/html/HTMLOptionElement.cpp:
```diff
@@ -93,6 +93,12 @@ bool HTMLOptionElement::isFocusable() const
+
     // Option elements do not have a renderer.
     auto* style = const_cast<HTMLOptionElement&>(*this).computedStyle();
     return style && style->display() != DisplayType::None;
```
The code below is now no longer executed.

Ahmad-S792 replied:
Should I move this to separate function: `HTMLOptGroup::isNotFocusableInMenuList` and rather than using `(select && select->)`, just use `(select->usesMenuList)`?

Ahmad-S792 replied:
Also in return, rather than returning 'HTMLElement::isFocusable', I return 'HTMLOptionElement::isFocusable'?

Ahmad-S792 replied:
Done. Removed and simplified.

nt1m commented on Source/WebCore/html/HTMLOptGroupElement.cpp:
```diff
@@ -65,6 +65,12 @@ bool HTMLOptGroupElement::isFocusable() const
+
     // Optgroup elements do not have a renderer.
     auto* style = const_cast<HTMLOptGroupElement&>(*this).computedStyle();
     return style && style->display() != DisplayType::None;
```
The code below is no longer executed

Ahmad-S792 replied:
Similar suggestion as above? Will it work?

nt1m replied:
I think the whole function can probably be:
```
     RefPtr select = ownerSelectElement();
     if (select && select->usesMenuList())
         return false;
     return HTMLElement::isFocusable();
```

The display & the supportsFocus() checks are both covered by `HTMLElement::isFocusable()`.

@rniwa Wdyt?

rniwa replied:
Yeah, it seems like we can just get rid of the code.

Ahmad-S792 replied:
Removed. :-)

nt1m reviewed (approved):
I’d change the commit title to remove the “is hidden” part since we were already checking that before.

Ahmad-S792 committed:
`option` and `optgroup` shouldn't be focusable in a menulist

https://bugs.webkit.org/show_bug.cgi?id=249627
rdar://problem/103717395

Reviewed by Tim Nguyen.

This patch aligns WebKit with Blink / Chromium and Gecko / Firefox.

Merge: https://chromium.googlesource.com/chromium/blink/+/db4eb7807f1591c3beebe2e40a3c57d9e57a3c85

This patch modifies WebKit behavior to match other browsers,
where `option` and `optgroup` elements are not focusable at all
when in menulist within `select` element irrespective of 'tabindex'
attribute and simplify code as well.

* Source/WebCore/html/HTMLOptionElement.cpp:
(HTMLOptionElement::isFocusable): return 'false' when in menulist
* Source/WebCore/html/HTMLOptGroupElement.cpp:
(HTMLOptGroupElement::isFocusable): Ditto
* LayoutTests/fast/forms/select/menulist-focusable-items.html: Add Test Case
* LayoutTests/fast/forms/select/menulist-focusable-items-expected.txt: Add Test Case Expectation
* LayoutTests/fast/forms/select/listbox-focusable-items.html: Add Test Case
* LayoutTests/fast/forms/select/listbox-focusable-items-expected.txt: Add Test Case Expectation
* LayoutTests/fast/events/resources/tabindex-focus-blur-all-iframe1.html: Rebaselined
* LayoutTests/fast/events/resources/tabindex-focus-blur-all-iframe2.html: Rebaselined
* LayoutTests/fast/events/resources/tabindex-focus-blur-all-frame1.html: Rebaselined
* LayoutTests/fast/events/resources/tabindex-focus-blur-all-frame2.html: Rebaselined
* LayoutTests/platform/ios/TestExpectations: Add 'ios' Expectation for 'fast/forms/select/listbox-focusable-items.html'

Canonical link: https://commits.webkit.org/264755@main
